### PR TITLE
[YUNIKORN-2123] Modify json.Unmarshal error message

### PR DESCRIPTION
**Pull request by makinyemi**
### What is this PR for?
Modifying the json.Unmarshal error message to better explain the failure of the specific function.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?

https://issues.apache.org/jira/browse/YUNIKORN-2123

### Screenshots (if appropriate)
n/a
### Questions:
n/a

### CODE REVIEW

**Comment by steinsgateted** [comment]
@makinyemi Thank you for the PR.
I found that there are a lot of "failed to unmarshal..."
Maybe we could make unified changes.

**Comment by makinyemi** [comment]
> @makinyemi Thank you for the PR. I found that there are a lot of "failed to unmarshal..." Maybe we could make unified changes.

@steinsgateted No problem! Ok, were you thinking unified in the sense of adding some sort of reusable function or constant string that we could format as needed to substitute for all of the occurrences of "failed to unmarshal..."?

**Comment by manirajv06** [comment]
> > I found that there are a lot of "failed to unmarshal..." Maybe we could make unified changes.
> 
> @steinsgateted No problem! Ok, were you thinking unified in the sense of adding some sort of reusable function or constant string that we could format as needed to substitute for all of the occurrences of "failed to unmarshal..."?

Introduce a const something like this.
STRING1 = "failed to unmarshal DAONAME response from response body"
replace DAONAME with actual name.

Also you can introduce constants for 

"Incorrect Status code"
"JSON error message is incorrect"

**Review by craigcondit — CHANGES_REQUESTED** [review]
-1. Every one of of the new messages is incorrect. It's the YAPIError structure that we're attempting to parse (Contains StatusCode, Message, Description). All of these functions are used to test error handling, and we assert in each of them that the *error object* is parsed correctly. Just change every message to "failed to unmarshal error response from response body".

**Review by craigcondit — CHANGES_REQUESTED** [review]
Now you've got additional changes that are *not* error handling. You need to pay attention to the context for each one.

**Comment by makinyemi** [comment]
> Now you've got additional changes that are _not_ error handling. You need to pay attention to the context for each one.

Thanks for the feedback. Will reset and go back through

**Review by craigcondit — APPROVED** [review]
+1, pending precommit checks.

**Comment by steinsgateted** [comment]
@makinyemi @manirajv06 @craigcondit 
Thank you for your discussion.
nit: I think other "failed to unmarshal" information can also be changed.

**Comment by makinyemi** [comment]
> nit: I think other "failed to unmarshal" information can also be changed.

Similar to my last commit? Make a constant for specific test functions?

File: pkg/webservice/handlers_test.go [hunk 1]
```diff
@@ -48,6 +48,8 @@ import (
 	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
 )
 
+const UnmarshalError = "Failed to unmarshal error response from response body"
```
Comment (steinsgateted) [inline pkg/webservice/handlers_test.go:51]: UnmarshalError -> unmarshalError

File: pkg/webservice/handlers_test.go [hunk 2]
```diff
@@ -321,7 +323,7 @@ func TestValidateConf(t *testing.T) {
 		validateConf(resp, req)
 		var vcr dao.ValidateConfResponse
 		err := json.Unmarshal(resp.outputBytes, &vcr)
-		assert.NilError(t, err, "failed to unmarshal ValidateConfResponse from response body")
+		assert.NilError(t, err, "%s (ValidateConfResponse)", UnmarshalError)
```
Comment (steinsgateted) [inline pkg/webservice/handlers_test.go:326]: dao can be removed
`assert.NilError(t, err, unmarshalError)`

**Review by steinsgateted — COMMENTED** [review]
> Just change every message to "failed to unmarshal error response from response body".

> Similar to my last commit? Make a constant for specific test functions?

@makinyemi @craigcondit 
Thank you.
I think it's like @craigcondit said.
Replace all messages with unmarshalError.
No need to retain the original DAONAME part.

**Comment by makinyemi** [comment]
> Replace all messages with unmarshalError. No need to retain the original DAONAME part.

@steinsgateted I did something like that in a previous commit that i reverted where @craigcondit mentioned in the above comment

> Now you've got additional changes that are _not_ error handling. You need to pay attention to the context for each one.

If i understand correctly, as @craigcondit explained it certain test were testing for specific failure and the messages reflected the error they were handling.

If i am mistaken i can definitely make the corrections

**Comment by makinyemi** [comment]
@steinsgateted @craigcondit Actually, I remember now. I added DAOInfo to every message, which was wrong. 
So if i understand correctly. We are good to remove all DAOInfo?

**Comment by steinsgateted** [comment]
@makinyemi Thank you
I think we can use two variables to deal with different situations.
Create a new variable:
`const unmarshalDAOError = "Failed to unmarshal dao response from response body"`
In test error handling situation:
`assert.NilError(t, err, unmarshalError)`
in dao:
`assert.NilError(t, err, "%s (ValidateConfResponse)", unmarshalDAOError)`

> Also you can introduce constants for
> 
> "Incorrect Status code"
> "JSON error message is incorrect"

we could add two new consts:
```
const statusCodeError = "Incorrect Status code"
const JsonMessageError = "JSON error message is incorrect"
```

**Comment by steinsgateted** [comment]
@makinyemi @manirajv06 @craigcondit Thank you.
I would like some clarifications.
I think there are three expressions of unmarshal dao messages.
Which expression method do you think makes more sense?

- Using the original method is like:

`assert.NilError(t, err, "failed to unmarshal ValidateConfResponse from response body")`

- `assert.NilError(t, err, "%s (ValidateConfResponse)", unmarshalDAOError)`

- `assert.NilError(t, err, unmarshalDAOError)`

**Comment by craigcondit** [comment]
> @makinyemi @manirajv06 @craigcondit Thank you. I would like some clarifications. I think there are three expressions of unmarshal dao messages. Which expression method do you think makes more sense?
> 
> * Using the original method is like:
> 
> `assert.NilError(t, err, "failed to unmarshal ValidateConfResponse from response body")`
> 
> * `assert.NilError(t, err, "%s (ValidateConfResponse)", unmarshalDAOError)`
> * `assert.NilError(t, err, unmarshalDAOError)`

What I think is that *far* too much time and effort has been spent on this. We're talking about the wording of a diagnostic message echoed by test cases when they fail (which they basically never do). Pick something reasonable and go with it.

**Comment by steinsgateted** [comment]
@makinyemi @craigcondit Thank you. +1

**Review by steinsgateted — APPROVED** [review]

**Review by pbacsko — APPROVED** [review]
+1

**Pull request closed without merging**